### ISSUE TITLE
mnist_siamese example

### ISSUE THREAD

**Issue by up201007037** [keras-team/keras#9269]
Hi,

I have two questions regarding the following: https://github.com/keras-team/keras/blob/master/examples/mnist_siamese.py

1) In the function `euclidean_distance` why do you use K.epislon? Because you are doing a square of a difference. The result will always be positive. 

2) In the paper that you cite, the labels are 0 when the two images are similar and 1 when they are dissimilar. The values of the loss function are not the same when you change that (in your case y=1 denotes similar images).

Thank you so much! Any help would be appreciated. 
@fchollet @ozabluda

**Comment by ozabluda** [comment]
1. Derivative `d/dx sqrt(x)` is infinite at x=0, without epsilon, you'll have NaNs during backprop. Also, due to numeric precision, without epsilon you may have negative number under sqrt(), although TF and TH take care about it automatically for you, don't know about CNTK.

2. I am not the initial author of that example, but labels going like `1: same, 0: different` are more consistent with Boolean logic and other [almost] drop-in replacement losses (cosine similarity followed by cross-entropy), etc), which are easier to try this way. Also, this way thresholding with 0.5 can be done by rounding. The biggest problem with labels like this IMHO, is that they are less consistent with MSE (which nobody wants anyway).

**Comment by up201007037** [comment]
Thank you so much @ozabluda for your answer. The first question, yes, it makes sense what you have said. I should have seen it, sorry for waste your time.
Regarding the second one, my question was that if you have two dissimilar images (according to the paper, y=1) and "LS is the partial loss function for a pair of similar points, LD the
partial loss function for a pair of dissimilar points". 
I saw the implementation again, you have changed the original function so that the LS  is the partial loss function for a pair of dissimilar points and LD is the partial loss function for a pair of similar points. You have switched the functions, so it is ok :) 

I just have doubts about the margin. It is not trivial to choose that. 
Thank you once again for your attention

**Comment by ozabluda** [comment]
>you have changed the original function so that the LS is the partial loss function for a pair of dissimilar points and LD is the partial loss function for a pair of similar points.

Do you have PR number? I don't remember doing it, and can't find it now.

**Comment by up201007037** [comment]
The function from the paper is not the same as the function that you use, but the function that you use is ok (because, for you Y=1 means similar). You did not change that, you implemented it like that since the beginning. 
Thank you a lot for the explanation about the margin :) 

Best
Carlos

**Comment by ozabluda** [comment]
I didn't implement it like that from the beginning. It was like that when I got to it. I didn't change it either, but I did fix it. Revision history is a bit confusing, because I renamed the file, but if you follow the following, it is more clear:

https://github.com/keras-team/keras/pulls?q=is%3Apr+author%3Aozabluda+siamese

**Comment by up201007037** [comment]
I see it. Thank you so much for your attention! :)
Best
